### PR TITLE
Add tests for main loop interruption and error handling

### PR DESCRIPTION
**Pull request by bartvm**
Note that part of this won't count towards test coverage because `coverage.py` doesn't recognise code run by another process.

### CODE REVIEW

**Comment by bartvm** [comment]
I don't know why it was there, but for some reason `format_exc(exception)` was used, but the argument to `format_exc` should be the traceback limit (an integer)... In Python 2 this silently fails, but in Python 3 it actually crashes, which happened with the unit test. (Long live unit tests.)

**Comment by bartvm** [comment]
#706, #707 and #708 are mostly just trying to cover some gaps in coverage. Could someone have a quick look and give the LGTM? @rizar @vdumoulin @dmitriy-serdyuk

File: tests/test_main_loop.py [hunk 1]
```diff
@@ -77,3 +64,41 @@ def do_test(with_serialization):
...
+    p.join()
+
+
+def test_error():
+    ext = TrainingExtension()
+    ext.after_batch = MagicMock(side_effect=KeyError)
+    ext.on_error = MagicMock()
+    main_loop = MockMainLoop(extensions=[ext, FinishAfter(after_epoch=True)])
+    assert_raises(KeyError, main_loop.run)
+    ext.on_error.assert_called_once_with()
```
Comment (dmitriy-serdyuk) [inline tests/test_main_loop.py]: It's possible to check that we have a log record `got_exception` here, isn't it?

Comment (bartvm) [inline tests/test_main_loop.py]: Yeah, I can add that.

Comment (bartvm) [inline tests/test_main_loop.py]: 6ed490590d04cafc03e4aa12729061083cdac384

**Comment by dmitriy-serdyuk** [comment]
Looks good, let's wait until it passes tests.

**Pull request merged**